File: 2.brute-force/동규/1.spiral-matrix.js
function spiralOrder(matrix) {
  let top = 0;
  let left = 0;
  let bottom = matrix.length - 1;
  let rigth = matrix[0].length - 1;
  const result = [];
  const size = matrix.length * matrix[0].length;

  while (result.length < size) {
    for (let i = left; i <= rigth && result.length < size; i++) {
      result.push(matrix[top][i]);
    }
    top++;
    for (let i = top; i <= bottom && result.length < size; i++) {
      result.push(matrix[i][rigth]);
    }
    rigth--;
    for (let i = rigth; i >= left && result.length < size; i--) {
      result.push(matrix[bottom][i]);
    }
    bottom--;
    for (let i = bottom; i >= top && result.length < size; i--) {
      result.push(matrix[i][left]);
    }
    left++;
  }
  return result;
}

File: 2.brute-force/지연/1.spiral-matrix.py
from typing import List


class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        result = []

        row_start_idx = 0
        row_end_idx = len(matrix) - 1     # 2

        col_start_idx = 0
        col_end_idx = len(matrix[0]) - 1  # 3

        while (row_start_idx <= row_end_idx and col_start_idx <= col_end_idx):
            for i in range(col_start_idx, col_end_idx + 1):        # 0, 4
                result.append(matrix[row_start_idx][i])            # 1, 2, 3, 4
            row_start_idx += 1                                     # 1

            for i in range(row_start_idx, row_end_idx + 1):        # 1, 3
                result.append(matrix[i][col_end_idx])              # 8, 12
            col_end_idx -= 1                                       # 2

            if (row_start_idx <= row_end_idx):                       # 1 <= 2
                for i in range(col_end_idx, col_start_idx - 1, -1):  # 2, 1, 0
                    result.append(matrix[row_end_idx][i])           # 11, 10, 9
                row_end_idx -= 1                                    # 1

            if (col_start_idx <= col_end_idx):                       # 0 <= 2
                for i in range(row_end_idx, row_start_idx - 1, -1):  # 1, 0
                    result.append(matrix[i][col_start_idx])          # 5
                col_start_idx += 1                                   # 1

        return result


print(Solution().spiralOrder([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]]))
